File: scripts/document_metadata.py
#!/usr/bin/env python3
"""Central document metadata rules for Coding Plugins docs."""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class DocumentArtifact:
    """Feature-scoped document artifact metadata contract."""

    label: str
    suffix: str
    directory: str
    relation_key: str
    doc_id_required: bool = True
    sync_upstream: tuple[str, ...] = ()


@dataclass(frozen=True)
class Frontmatter:
    """Parsed Markdown frontmatter with stable render order."""

    scalars: dict[str, str]
    lists: dict[str, list[str]]
    order: list[str]


FEATURE_README_METADATA_REQUIRED_FIELDS = ("title", "status", "feature", "updated")
EVIDENCE_METADATA_REQUIRED_FIELDS = ("title", "status", "feature", "created", "updated")
ARCHIVED_EVIDENCE_METADATA_REQUIRED_FIELDS = (
    "title",
    "status",
    "feature",
    "created",
    "updated",
    "validation_mode",
    "archive_of",
    "archived_at",
)
PLAN_METADATA_REQUIRED_FIELDS = ("title", "status", "feature", "created", "updated")

DOCUMENT_ARTIFACTS = (
    DocumentArtifact("PRD", "PRD", "requirements", "related_specs"),
    DocumentArtifact("TDD", "TDD", "technicals", "related_technical", sync_upstream=("PRD",)),
    DocumentArtifact("TID", "TID", "technicals", "related_technical", sync_upstream=("PRD", "TDD")),
    DocumentArtifact("TCD", "TCD", "test-cases", "related_test_cases", sync_upstream=("PRD", "TDD", "TID")),
    DocumentArtifact("IPD", "IPD", "plans", "related_plans", sync_upstream=("PRD", "TDD", "TID", "TCD")),
    DocumentArtifact("TED", "TED", "evidences", "related_evidence", sync_upstream=("PRD", "TDD", "TID", "TCD", "IPD")),
)
ARTIFACTS_BY_SUFFIX = {artifact.suffix: artifact for artifact in DOCUMENT_ARTIFACTS}
ARTIFACT_SUFFIXES = tuple(artifact.suffix for artifact in DOCUMENT_ARTIFACTS)
RELATION_KEYS = tuple(dict.fromkeys(artifact.relation_key for artifact in DOCUMENT_ARTIFACTS))
DOCUMENT_SYNC_DEPENDENCIES = {
    artifact.suffix: artifact.sync_upstream for artifact in DOCUMENT_ARTIFACTS if artifact.sync_upstream
}


def split_frontmatter(text: str) -> tuple[list[str], str]:
    if not text.startswith("---\n"):
        return [], text
    end = text.find("\n---", 4)
    if end == -1:
        return [], text

    return text[4:end].splitlines(), text[end + len("\n---") :].lstrip("\n")


def parse_frontmatter_block(lines: list[str]) -> Frontmatter:
    scalars: dict[str, str] = {}
    lists: dict[str, list[str]] = {}
    order: list[str] = []
    current_key: str | None = None

    for line in lines:
        stripped = line.strip()
        if not stripped:
            continue
        if line.startswith(" ") and current_key and stripped.startswith("- "):
            lists.setdefault(current_key, []).append(stripped[2:].strip().strip('"').strip("'"))
            continue
        if ":" not in line or line.startswith(" "):
            continue
        key, value = line.split(":", 1)
        current_key = key.strip()
        if current_key not in order:
            order.append(current_key)

        raw_value = value.strip()
        value = raw_value.strip('"').strip("'")
        if raw_value == "[]":
            lists.setdefault(current_key, [])
        elif value:
            scalars[current_key] = value
        else:
            lists.setdefault(current_key, [])

    return Frontmatter(scalars=scalars, lists=lists, order=order)


def render_frontmatter_block(frontmatter: Frontmatter) -> str:
    keys = list(frontmatter.order)
    for key in list(frontmatter.scalars) + list(frontmatter.lists):
        if key not in keys:
            keys.append(key)

    lines = ["---"]
    for key in keys:
        if key in frontmatter.lists:
            values = frontmatter.lists[key]
            if values:
                lines.append(f"{key}:")
                lines.extend(f"  - {value}" for value in values)
            else:
                lines.append(f"{key}: []")
        elif key in frontmatter.scalars:
            lines.append(f"{key}: {frontmatter.scalars[key]}")
    lines.append("---")
    return "\n".join(lines) + "\n"


def parse_frontmatter(text: str) -> dict[str, str]:
    lines, _body = split_frontmatter(text)
    if not lines:
        return {}
    return dict(parse_frontmatter_block(lines).scalars)


def frontmatter_list_values(text: str, key: str) -> list[str]:
    lines, _body = split_frontmatter(text)
    if not lines:
        return []
    return list(parse_frontmatter_block(lines).lists.get(key, []))


def feature_docs_root(root: Path) -> Path:
    return root / "docs" / "coding-plugins" / "features"


def collect_feature_roots(root: Path) -> list[Path]:
    features_root = feature_docs_root(root)
    if not features_root.exists():
        return []

    return sorted(path for path in features_root.iterdir() if path.is_dir())


def feature_root_for_document(root: Path, path: Path) -> tuple[str, Path] | None:
    try:
        relative_parts = path.relative_to(feature_docs_root(root)).parts
    except ValueError:
        return None
    if len(relative_parts) < 1:
        return None
    feature = relative_parts[0]
    return feature, feature_docs_root(root) / feature


def artifact_for_suffix(suffix: str) -> DocumentArtifact:
    try:
        return ARTIFACTS_BY_SUFFIX[suffix]
    except KeyError as error:
        raise ValueError(f"Unknown document artifact suffix: {suffix}") from error


def artifact_directories() -> tuple[str, ...]:
    return tuple(dict.fromkeys(artifact.directory for artifact in DOCUMENT_ARTIFACTS))


def filename_patterns_by_directory() -> dict[str, re.Pattern[str]]:
    suffixes_by_directory: dict[str, list[str]] = {}
    for artifact in DOCUMENT_ARTIFACTS:
        suffixes_by_directory.setdefault(artifact.directory, []).append(artifact.suffix)

    return {
        directory: re.compile(r"^[A-Za-z0-9_.-]+-(?:" + "|".join(map(re.escape, suffixes)) + r")\.md$")
        for directory, suffixes in suffixes_by_directory.items()
    }


def document_doc_id(path: Path) -> str:
    for suffix in ARTIFACT_SUFFIXES:
        marker = f"-{suffix}"
        if path.stem.endswith(marker):
            return path.stem[: -len(marker)]
    return path.stem


def document_suffix(path: Path) -> str | None:
    for suffix in ARTIFACT_SUFFIXES:
        if path.stem.endswith(f"-{suffix}"):
            return suffix
    return None


def feature_artifact_file(feature_root: Path, directory: str, suffix: str, doc_id: str | None = None) -> Path:
    return feature_root / directory / f"{doc_id or feature_root.name}-{suffix}.md"


def artifact_file(feature_root: Path, suffix: str, doc_id: str | None = None) -> Path:
    artifact = artifact_for_suffix(suffix)
    return feature_artifact_file(feature_root, artifact.directory, artifact.suffix, doc_id)


def feature_artifact_files(feature_root: Path, directory: str, suffix: str) -> list[Path]:
    artifact_dir = feature_root / directory
    if not artifact_dir.exists():
        return []
    return sorted(path for path in artifact_dir.glob(f"*-{suffix}.md") if path.is_file())


def artifact_files(feature_root: Path, suffix: str) -> list[Path]:
    artifact = artifact_for_suffix(suffix)
    return feature_artifact_files(feature_root, artifact.directory, artifact.suffix)


def feature_artifact_files_for_doc_id(feature_root: Path, directory: str, suffix: str, doc_id: str) -> list[Path]:
    path = feature_artifact_file(feature_root, directory, suffix, doc_id)
    return [path] if path.exists() else []


def artifact_files_for_doc_id(feature_root: Path, suffix: str, doc_id: str) -> list[Path]:
    artifact = artifact_for_suffix(suffix)
    return feature_artifact_files_for_doc_id(feature_root, artifact.directory, artifact.suffix, doc_id)


def feature_doc_ids(feature_root: Path) -> list[str]:
    doc_ids = {document_doc_id(path) for artifact in DOCUMENT_ARTIFACTS for path in artifact_files(feature_root, artifact.suffix)}
    return sorted(doc_ids)


def documents_by_suffix_for_doc_id(feature_root: Path, doc_id: str) -> dict[str, list[Path]]:
    return {artifact.suffix: artifact_files_for_doc_id(feature_root, artifact.suffix, doc_id) for artifact in DOCUMENT_ARTIFACTS}


def expected_related_paths_for_doc_id(
    feature_root: Path,
    doc_id: str,
    source_path: Path | None = None,
) -> dict[str, list[Path]]:
    source = source_path.resolve() if source_path is not None else None
    expected: dict[str, list[Path]] = {key: [] for key in RELATION_KEYS}

    for artifact in DOCUMENT_ARTIFACTS:
        for path in artifact_files_for_doc_id(feature_root, artifact.suffix, doc_id):
            if source is not None and path.resolve() == source:
                continue
            expected[artifact.relation_key].append(path)

    return {key: sorted(paths) for key, paths in expected.items()}

File: src/lib/document-metadata.ts
import { existsSync, readdirSync } from "node:fs";
import { basename, dirname, join, relative, resolve } from "node:path";

export interface DocumentArtifact {
  label: string;
  suffix: string;
  directory: string;
  relationKey: string;
  legacyRelationKey: string;
  docIdRequired: boolean;
  syncUpstream: string[];
}

export interface Frontmatter {
  scalars: Record<string, string>;
  lists: Record<string, string[]>;
  order: string[];
}

export const FEATURE_README_METADATA_REQUIRED_FIELDS = ["title", "status", "feature", "updated"];
export const EVIDENCE_METADATA_REQUIRED_FIELDS = ["title", "status", "feature", "created", "updated"];
export const ARCHIVED_EVIDENCE_METADATA_REQUIRED_FIELDS = [
  "title",
  "status",
  "feature",
  "created",
  "updated",
  "validation_mode",
  "archive_of",
  "archived_at",
];
export const PLAN_METADATA_REQUIRED_FIELDS = ["title", "status", "feature", "created", "updated"];

export const RELATED_DOCS_KEY = "related_docs";
export const LEGACY_RELATION_KEYS = ["related_specs", "related_technical", "related_test_cases", "related_plans", "related_evidence"];

export const DOCUMENT_ARTIFACTS: DocumentArtifact[] = [
  { label: "PRD", suffix: "PRD", directory: "requirements", relationKey: RELATED_DOCS_KEY, legacyRelationKey: "related_specs", docIdRequired: true, syncUpstream: [] },
  { label: "TDD", suffix: "TDD", directory: "technicals", relationKey: RELATED_DOCS_KEY, legacyRelationKey: "related_technical", docIdRequired: true, syncUpstream: ["PRD"] },
  { label: "TID", suffix: "TID", directory: "technicals", relationKey: RELATED_DOCS_KEY, legacyRelationKey: "related_technical", docIdRequired: true, syncUpstream: ["PRD", "TDD"] },
  { label: "TCD", suffix: "TCD", directory: "test-cases", relationKey: RELATED_DOCS_KEY, legacyRelationKey: "related_test_cases", docIdRequired: true, syncUpstream: ["PRD", "TDD", "TID"] },
  { label: "IPD", suffix: "IPD", directory: "plans", relationKey: RELATED_DOCS_KEY, legacyRelationKey: "related_plans", docIdRequired: true, syncUpstream: ["PRD", "TDD", "TID", "TCD"] },
  { label: "TED", suffix: "TED", directory: "evidences", relationKey: RELATED_DOCS_KEY, legacyRelationKey: "related_evidence", docIdRequired: true, syncUpstream: ["PRD", "TDD", "TID", "TCD", "IPD"] },
];

const ARTIFACTS_BY_SUFFIX = new Map(DOCUMENT_ARTIFACTS.map((artifact) => [artifact.suffix, artifact]));
export const ARTIFACT_SUFFIXES = DOCUMENT_ARTIFACTS.map((artifact) => artifact.suffix);
export const RELATION_KEYS = [...new Set(DOCUMENT_ARTIFACTS.map((artifact) => artifact.relationKey))];
export const DOCUMENT_SYNC_DEPENDENCIES = Object.fromEntries(
  DOCUMENT_ARTIFACTS.filter((artifact) => artifact.syncUpstream.length > 0).map((artifact) => [artifact.suffix, artifact.syncUpstream]),
);

export function splitFrontmatter(text: string): [string[], string] {
  if (!text.startsWith("---\n")) {
    return [[], text];
  }
  const end = text.indexOf("\n---", 4);
  if (end === -1) {
    return [[], text];
  }
  return [text.slice(4, end).split(/\r?\n/), text.slice(end + "\n---".length).replace(/^\n/, "")];
}

export function parseFrontmatterBlock(lines: string[]): Frontmatter {
  const scalars: Record<string, string> = {};
  const lists: Record<string, string[]> = {};
  const order: string[] = [];
  let currentKey: string | undefined;

  for (const line of lines) {
    const stripped = line.trim();
    if (!stripped) {
      continue;
    }
    if (line.startsWith(" ") && currentKey && stripped.startsWith("- ")) {
      lists[currentKey] ??= [];
      lists[currentKey].push(stripped.slice(2).trim().replace(/^["']|["']$/g, ""));
      continue;
    }
    if (!line.includes(":") || line.startsWith(" ")) {
      continue;
    }
    const index = line.indexOf(":");
    currentKey = line.slice(0, index).trim();
    if (!order.includes(currentKey)) {
      order.push(currentKey);
    }
    const rawValue = line.slice(index + 1).trim();
    const value = rawValue.replace(/^["']|["']$/g, "");
    if (rawValue === "[]") {
      lists[currentKey] ??= [];
    } else if (value) {
      scalars[currentKey] = value;
    } else {
      lists[currentKey] ??= [];
    }
  }

  return { scalars, lists, order };
}

export function renderFrontmatterBlock(frontmatter: Frontmatter): string {
  const keys = [...frontmatter.order];
  for (const key of [...Object.keys(frontmatter.scalars), ...Object.keys(frontmatter.lists)]) {
    if (!keys.includes(key)) {
      keys.push(key);
    }
  }

  const lines = ["---"];
  for (const key of keys) {
    if (key in frontmatter.lists) {
      const values = frontmatter.lists[key];
      if (values.length > 0) {
        lines.push(`${key}:`);
        lines.push(...values.map((value) => `  - ${value}`));
      } else {
        lines.push(`${key}: []`);
      }
    } else if (key in frontmatter.scalars) {
      lines.push(`${key}: ${frontmatter.scalars[key]}`);
    }
  }
  lines.push("---");
  return `${lines.join("\n")}\n`;
}

export function parseFrontmatter(text: string): Record<string, string> {
  const [lines] = splitFrontmatter(text);
  if (lines.length === 0) {
    return {};
  }
  return { ...parseFrontmatterBlock(lines).scalars };
}

export function frontmatterListValues(text: string, key: string): string[] {
  const [lines] = splitFrontmatter(text);
  if (lines.length === 0) {
    return [];
  }
  return [...(parseFrontmatterBlock(lines).lists[key] ?? [])];
}

export function featureDocsRoot(root: string): string {
  return join(root, "docs/coding-plugins/features");
}

export function collectFeatureRoots(root: string): string[] {
  const featuresRoot = featureDocsRoot(root);
  if (!existsSync(featuresRoot)) {
    return [];
  }
  return readdirSync(featuresRoot, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => join(featuresRoot, entry.name))
    .sort();
}

export function featureRootForDocument(root: string, path: string): [string, string] | undefined {
  const featuresRoot = featureDocsRoot(root);
  const rel = relative(featuresRoot, path);
  if (rel.startsWith("..") || rel === "") {
    return undefined;
  }
  const [feature] = rel.split(/[\\/]/);
  if (!feature) {
    return undefined;
  }
  return [feature, join(featuresRoot, feature)];
}

export function artifactForSuffix(suffix: string): DocumentArtifact {
  const artifact = ARTIFACTS_BY_SUFFIX.get(suffix);
  if (!artifact) {
    throw new Error(`Unknown document artifact suffix: ${suffix}`);
  }
  return artifact;
}

export function artifactDirectories(): string[] {
  return [...new Set(DOCUMENT_ARTIFACTS.map((artifact) => artifact.directory))];
}

export function filenamePatternsByDirectory(): Record<string, RegExp> {
  const suffixesByDirectory: Record<string, string[]> = {};
  for (const artifact of DOCUMENT_ARTIFACTS) {
    suffixesByDirectory[artifact.directory] ??= [];
    suffixesByDirectory[artifact.directory].push(artifact.suffix);
  }
  return Object.fromEntries(
    Object.entries(suffixesByDirectory).map(([directoryName, suffixes]) => [
      directoryName,
      new RegExp(`^[A-Za-z0-9_.-]+-(?:${suffixes.map(escapeRegExp).join("|")})\\.md$`),
    ]),
  );
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function documentDocId(path: string): string {
  const stem = basename(path).replace(/\.[^.]+$/, "");
  for (const suffix of ARTIFACT_SUFFIXES) {
    const marker = `-${suffix}`;
    if (stem.endsWith(marker)) {
      return stem.slice(0, -marker.length);
    }
  }
  return stem;
}

export function documentSuffix(path: string): string | undefined {
  const stem = basename(path).replace(/\.[^.]+$/, "");
  return ARTIFACT_SUFFIXES.find((suffix) => stem.endsWith(`-${suffix}`));
}

export function featureArtifactFile(featureRoot: string, directoryName: string, suffix: string, docId?: string): string {
  return join(featureRoot, directoryName, `${docId ?? basename(featureRoot)}-${suffix}.md`);
}

export function artifactFile(featureRoot: string, suffix: string, docId?: string): string {
  const artifact = artifactForSuffix(suffix);
  return featureArtifactFile(featureRoot, artifact.directory, artifact.suffix, docId);
}

export function featureArtifactFiles(featureRoot: string, directoryName: string, suffix: string): string[] {
  const artifactDir = join(featureRoot, directoryName);
  if (!existsSync(artifactDir)) {
    return [];
  }
  return readdirSync(artifactDir, { withFileTypes: true })
    .filter((entry) => entry.isFile() && entry.name.endsWith(`-${suffix}.md`))
    .map((entry) => join(artifactDir, entry.name))
    .sort();
}

export function artifactFiles(featureRoot: string, suffix: string): string[] {
  const artifact = artifactForSuffix(suffix);
  return featureArtifactFiles(featureRoot, artifact.directory, artifact.suffix);
}

export function featureArtifactFilesForDocId(featureRoot: string, directoryName: string, suffix: string, docId: string): string[] {
  const path = featureArtifactFile(featureRoot, directoryName, suffix, docId);
  return existsSync(path) ? [path] : [];
}

export function artifactFilesForDocId(featureRoot: string, suffix: string, docId: string): string[] {
  const artifact = artifactForSuffix(suffix);
  return featureArtifactFilesForDocId(featureRoot, artifact.directory, artifact.suffix, docId);
}

export function featureDocIds(featureRoot: string): string[] {
  const docIds = new Set<string>();
  for (const artifact of DOCUMENT_ARTIFACTS) {
    for (const path of artifactFiles(featureRoot, artifact.suffix)) {
      docIds.add(documentDocId(path));
    }
  }
  return [...docIds].sort();
}

export function documentsBySuffixForDocId(featureRoot: string, docId: string): Record<string, string[]> {
  return Object.fromEntries(DOCUMENT_ARTIFACTS.map((artifact) => [artifact.suffix, artifactFilesForDocId(featureRoot, artifact.suffix, docId)]));
}

export function expectedRelatedPathsForDocId(featureRoot: string, docId: string, sourcePath?: string): Record<string, string[]> {
  const source = sourcePath ? resolve(sourcePath) : undefined;
  const expected: Record<string, string[]> = Object.fromEntries(RELATION_KEYS.map((key) => [key, []]));
  for (const artifact of DOCUMENT_ARTIFACTS) {
    for (const path of artifactFilesForDocId(featureRoot, artifact.suffix, docId)) {
      if (source && resolve(path) === source) {
        continue;
      }
      expected[artifact.relationKey].push(path);
    }
  }
  for (const key of Object.keys(expected)) {
    expected[key].sort();
  }
  return expected;
}

export function featureName(featureRoot: string): string {
  return basename(featureRoot);
}

export function parentDir(path: string): string {
  return dirname(path);
}
